File: basic_dsa_two_hundred_thirty_one.java
/**
 * Basic DSA Problems - Is Subsequence
 * Check if s is subsequence of t (relative order preserved)
 */
public class basic_dsa_two_hundred_thirty_one {

    // Problem: s and t. Return true if s is subsequence of t.

    /**
     * Two pointers: match s[i] with t[j]; advance j always, i when match. True if i reached end of s.
     * Time O(|t|), Space O(1)
     */
    public static boolean isSubsequence(String s, String t) {
        int i = 0, j = 0;
        while (i < s.length() && j < t.length()) {
            if (s.charAt(i) == t.charAt(j)) i++;
            j++;
        }
        return i == s.length();
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Is Subsequence");
        System.out.println("=".repeat(60));

        System.out.println("abc, ahbgdc -> " + isSubsequence("abc", "ahbgdc"));
        System.out.println("axc, ahbgdc -> " + isSubsequence("axc", "ahbgdc"));
    }
}

File: basic_dsa_two_hundred_thirty_one.py
"""
Basic DSA - Longest Common Subsequence
Given two strings text1 and text2, return the length of their longest common subsequence.
"""


def longest_common_subsequence(text1, text2):
    """
    DP with rolling row.
    Time Complexity: O(n*m)
    Space Complexity: O(min(n, m))
    """
    if len(text2) > len(text1):
        text1, text2 = text2, text1
    # text2 is shorter
    m = len(text2)
    dp = [0] * (m + 1)
    for ch1 in text1:
        prev_diag = 0
        for j, ch2 in enumerate(text2, start=1):
            tmp = dp[j]
            if ch1 == ch2:
                dp[j] = prev_diag + 1
            else:
                dp[j] = max(dp[j], dp[j - 1])
            prev_diag = tmp
    return dp[m]


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Longest Common Subsequence")
    print("=" * 60)

    print(longest_common_subsequence("abcde", "ace"))  # 3
    print(longest_common_subsequence("abc", "abc"))  # 3
    print(longest_common_subsequence("abc", "def"))  # 0
